// @bokeh/bokehjs/build/js/lib/models/dom/styles.js
import { Model } from "../../model";
export class Styles extends Model {
    static __name__ = "Styles";
    constructor(attrs) {
        super(attrs);
    }
    static {
        this.define(({ Str, Nullable }) => ({
            align_content: [Nullable(Str), null],
            align_items: [Nullable(Str), null],
            align_self: [Nullable(Str), null],
            alignment_baseline: [Nullable(Str), null],
            all: [Nullable(Str), null],
            animation: [Nullable(Str), null],
            animation_delay: [Nullable(Str), null],
            animation_direction: [Nullable(Str), null],
            animation_duration: [Nullable(Str), null],
            animation_fill_mode: [Nullable(Str), null],
            animation_iteration_count: [Nullable(Str), null],
            animation_name: [Nullable(Str), null],
            animation_play_state: [Nullable(Str), null],
            animation_timing_function: [Nullable(Str), null],
            aspect_ratio: [Nullable(Str), null],
            backface_visibility: [Nullable(Str), null],
            background: [Nullable(Str), null],
            background_attachment: [Nullable(Str), null],
            background_clip: [Nullable(Str), null],
            background_color: [Nullable(Str), null],
            background_image: [Nullable(Str), null],
            background_origin: [Nullable(Str), null],
            background_position: [Nullable(Str), null],
            background_position_x: [Nullable(Str), null],
            background_position_y: [Nullable(Str), null],
            background_repeat: [Nullable(Str), null],
            background_size: [Nullable(Str), null],
            baseline_shift: [Nullable(Str), null],
            block_size: [Nullable(Str), null],
            border: [Nullable(Str), null],
            border_block_end: [Nullable(Str), null],
            border_block_end_color: [Nullable(Str), null],
            border_block_end_style: [Nullable(Str), null],
            border_block_end_width: [Nullable(Str), null],
            border_block_start: [Nullable(Str), null],
            border_block_start_color: [Nullable(Str), null],
            border_block_start_style: [Nullable(Str), null],
            border_block_start_width: [Nullable(Str), null],
            border_bottom: [Nullable(Str), null],
            border_bottom_color: [Nullable(Str), null],
            border_bottom_left_radius: [Nullable(Str), null],
            border_bottom_right_radius: [Nullable(Str), null],
            border_bottom_style: [Nullable(Str), null],
            border_bottom_width: [Nullable(Str), null],
            border_collapse: [Nullable(Str), null],
            border_color: [Nullable(Str), null],
            border_image: [Nullable(Str), null],
            border_image_outset: [Nullable(Str), null],
            border_image_repeat: [Nullable(Str), null],
            border_image_slice: [Nullable(Str), null],
            border_image_source: [Nullable(Str), null],
            border_image_width: [Nullable(Str), null],
            border_inline_end: [Nullable(Str), null],
            border_inline_end_color: [Nullable(Str), null],
            border_inline_end_style: [Nullable(Str), null],
            border_inline_end_width: [Nullable(Str), null],
            border_inline_start: [Nullable(Str), null],
            border_inline_start_color: [Nullable(Str), null],
            border_inline_start_style: [Nullable(Str), null],
            border_inline_start_width: [Nullable(Str), null],
            border_left: [Nullable(Str), null],
            border_left_color: [Nullable(Str), null],
            border_left_style: [Nullable(Str), null],
            border_left_width: [Nullable(Str), null],
            border_radius: [Nullable(Str), null],
            border_right: [Nullable(Str), null],
            border_right_color: [Nullable(Str), null],
            border_right_style: [Nullable(Str), null],
            border_right_width: [Nullable(Str), null],
            border_spacing: [Nullable(Str), null],
            border_style: [Nullable(Str), null],
            border_top: [Nullable(Str), null],
            border_top_color: [Nullable(Str), null],
            border_top_left_radius: [Nullable(Str), null],
            border_top_right_radius: [Nullable(Str), null],
            border_top_style: [Nullable(Str), null],
            border_top_width: [Nullable(Str), null],
            border_width: [Nullable(Str), null],
            bottom: [Nullable(Str), null],
            box_shadow: [Nullable(Str), null],
            box_sizing: [Nullable(Str), null],
            break_after: [Nullable(Str), null],
            break_before: [Nullable(Str), null],
            break_inside: [Nullable(Str), null],
            caption_side: [Nullable(Str), null],
            caret_color: [Nullable(Str), null],
            clear: [Nullable(Str), null],
            clip: [Nullable(Str), null],
            clip_path: [Nullable(Str), null],
            clip_rule: [Nullable(Str), null],
            color: [Nullable(Str), null],
            color_interpolation: [Nullable(Str), null],
            color_interpolation_filters: [Nullable(Str), null],
            column_count: [Nullable(Str), null],
            column_fill: [Nullable(Str), null],
            column_gap: [Nullable(Str), null],
            column_rule: [Nullable(Str), null],
            column_rule_color: [Nullable(Str), null],
            column_rule_style: [Nullable(Str), null],
            column_rule_width: [Nullable(Str), null],
            column_span: [Nullable(Str), null],
            column_width: [Nullable(Str), null],
            columns: [Nullable(Str), null],
            content: [Nullable(Str), null],
            counter_increment: [Nullable(Str), null],
            counter_reset: [Nullable(Str), null],
            cursor: [Nullable(Str), null],
            direction: [Nullable(Str), null],
            display: [Nullable(Str), null],
            dominant_baseline: [Nullable(Str), null],
            empty_cells: [Nullable(Str), null],
            fill: [Nullable(Str), null],
            fill_opacity: [Nullable(Str), null],
            fill_rule: [Nullable(Str), null],
            filter: [Nullable(Str), null],
            flex: [Nullable(Str), null],
            flex_basis: [Nullable(Str), null],
            flex_direction: [Nullable(Str), null],
            flex_flow: [Nullable(Str), null],
            flex_grow: [Nullable(Str), null],
            flex_shrink: [Nullable(Str), null],
            flex_wrap: [Nullable(Str), null],
            float: [Nullable(Str), null],
            flood_color: [Nullable(Str), null],
            flood_opacity: [Nullable(Str), null],
            font: [Nullable(Str), null],
            font_family: [Nullable(Str), null],
            font_feature_settings: [Nullable(Str), null],
            font_kerning: [Nullable(Str), null],
            font_size: [Nullable(Str), null],
            font_size_adjust: [Nullable(Str), null],
            font_stretch: [Nullable(Str), null],
            font_style: [Nullable(Str), null],
            font_synthesis: [Nullable(Str), null],
            font_variant: [Nullable(Str), null],
            font_variant_caps: [Nullable(Str), null],
            font_variant_east_asian: [Nullable(Str), null],
            font_variant_ligatures: [Nullable(Str), null],
            font_variant_numeric: [Nullable(Str), null],
            font_variant_position: [Nullable(Str), null],
            font_weight: [Nullable(Str), null],
            gap: [Nullable(Str), null],
            glyph_orientation_vertical: [Nullable(Str), null],
            grid: [Nullable(Str), null],
            grid_area: [Nullable(Str), null],
            grid_auto_columns: [Nullable(Str), null],
            grid_auto_flow: [Nullable(Str), null],
            grid_auto_rows: [Nullable(Str), null],
            grid_column: [Nullable(Str), null],
            grid_column_end: [Nullable(Str), null],
            grid_column_gap: [Nullable(Str), null],
            grid_column_start: [Nullable(Str), null],
            grid_gap: [Nullable(Str), null],
            grid_row: [Nullable(Str), null],
            grid_row_end: [Nullable(Str), null],
            grid_row_gap: [Nullable(Str), null],
            grid_row_start: [Nullable(Str), null],
            grid_template: [Nullable(Str), null],
            grid_template_areas: [Nullable(Str), null],
            grid_template_columns: [Nullable(Str), null],
            grid_template_rows: [Nullable(Str), null],
            height: [Nullable(Str), null],
            hyphens: [Nullable(Str), null],
            image_orientation: [Nullable(Str), null],
            image_rendering: [Nullable(Str), null],
            inline_size: [Nullable(Str), null],
            justify_content: [Nullable(Str), null],
            justify_items: [Nullable(Str), null],
            justify_self: [Nullable(Str), null],
            left: [Nullable(Str), null],
            letter_spacing: [Nullable(Str), null],
            lighting_color: [Nullable(Str), null],
            line_break: [Nullable(Str), null],
            line_height: [Nullable(Str), null],
            list_style: [Nullable(Str), null],
            list_style_image: [Nullable(Str), null],
            list_style_position: [Nullable(Str), null],
            list_style_type: [Nullable(Str), null],
            margin: [Nullable(Str), null],
            margin_block_end: [Nullable(Str), null],
            margin_block_start: [Nullable(Str), null],
            margin_bottom: [Nullable(Str), null],
            margin_inline_end: [Nullable(Str), null],
            margin_inline_start: [Nullable(Str), null],
            margin_left: [Nullable(Str), null],
            margin_right: [Nullable(Str), null],
            margin_top: [Nullable(Str), null],
            marker: [Nullable(Str), null],
            marker_end: [Nullable(Str), null],
            marker_mid: [Nullable(Str), null],
            marker_start: [Nullable(Str), null],
            mask: [Nullable(Str), null],
            mask_composite: [Nullable(Str), null],
            mask_image: [Nullable(Str), null],
            mask_position: [Nullable(Str), null],
            mask_repeat: [Nullable(Str), null],
            mask_size: [Nullable(Str), null],
            mask_type: [Nullable(Str), null],
            max_block_size: [Nullable(Str), null],
            max_height: [Nullable(Str), null],
            max_inline_size: [Nullable(Str), null],
            max_width: [Nullable(Str), null],
            min_block_size: [Nullable(Str), null],
            min_height: [Nullable(Str), null],
            min_inline_size: [Nullable(Str), null],
            min_width: [Nullable(Str), null],
            object_fit: [Nullable(Str), null],
            object_position: [Nullable(Str), null],
            opacity: [Nullable(Str), null],
            order: [Nullable(Str), null],
            orphans: [Nullable(Str), null],
            outline: [Nullable(Str), null],
            outline_color: [Nullable(Str), null],
            outline_offset: [Nullable(Str), null],
            outline_style: [Nullable(Str), null],
            outline_width: [Nullable(Str), null],
            overflow: [Nullable(Str), null],
            overflow_anchor: [Nullable(Str), null],
            overflow_wrap: [Nullable(Str), null],
            overflow_x: [Nullable(Str), null],
            overflow_y: [Nullable(Str), null],
            overscroll_behavior: [Nullable(Str), null],
            overscroll_behavior_block: [Nullable(Str), null],
            overscroll_behavior_inline: [Nullable(Str), null],
            overscroll_behavior_x: [Nullable(Str), null],
            overscroll_behavior_y: [Nullable(Str), null],
            padding: [Nullable(Str), null],
            padding_block_end: [Nullable(Str), null],
            padding_block_start: [Nullable(Str), null],
            padding_bottom: [Nullable(Str), null],
            padding_inline_end: [Nullable(Str), null],
            padding_inline_start: [Nullable(Str), null],
            padding_left: [Nullable(Str), null],
            padding_right: [Nullable(Str), null],
            padding_top: [Nullable(Str), null],
            page_break_after: [Nullable(Str), null],
            page_break_before: [Nullable(Str), null],
            page_break_inside: [Nullable(Str), null],
            paint_order: [Nullable(Str), null],
            perspective: [Nullable(Str), null],
            perspective_origin: [Nullable(Str), null],
            place_content: [Nullable(Str), null],
            place_items: [Nullable(Str), null],
            place_self: [Nullable(Str), null],
            pointer_events: [Nullable(Str), null],
            position: [Nullable(Str), null],
            quotes: [Nullable(Str), null],
            resize: [Nullable(Str), null],
            right: [Nullable(Str), null],
            rotate: [Nullable(Str), null],
            row_gap: [Nullable(Str), null],
            ruby_align: [Nullable(Str), null],
            ruby_position: [Nullable(Str), null],
            scale: [Nullable(Str), null],
            scroll_behavior: [Nullable(Str), null],
            shape_rendering: [Nullable(Str), null],
            stop_color: [Nullable(Str), null],
            stop_opacity: [Nullable(Str), null],
            stroke: [Nullable(Str), null],
            stroke_dasharray: [Nullable(Str), null],
            stroke_dashoffset: [Nullable(Str), null],
            stroke_linecap: [Nullable(Str), null],
            stroke_linejoin: [Nullable(Str), null],
            stroke_miterlimit: [Nullable(Str), null],
            stroke_opacity: [Nullable(Str), null],
            stroke_width: [Nullable(Str), null],
            tab_size: [Nullable(Str), null],
            table_layout: [Nullable(Str), null],
            text_align: [Nullable(Str), null],
            text_align_last: [Nullable(Str), null],
            text_anchor: [Nullable(Str), null],
            text_combine_upright: [Nullable(Str), null],
            text_decoration: [Nullable(Str), null],
            text_decoration_color: [Nullable(Str), null],
            text_decoration_line: [Nullable(Str), null],
            text_decoration_style: [Nullable(Str), null],
            text_emphasis: [Nullable(Str), null],
            text_emphasis_color: [Nullable(Str), null],
            text_emphasis_position: [Nullable(Str), null],
            text_emphasis_style: [Nullable(Str), null],
            text_indent: [Nullable(Str), null],
            text_justify: [Nullable(Str), null],
            text_orientation: [Nullable(Str), null],
            text_overflow: [Nullable(Str), null],
            text_rendering: [Nullable(Str), null],
            text_shadow: [Nullable(Str), null],
            text_transform: [Nullable(Str), null],
            text_underline_position: [Nullable(Str), null],
            top: [Nullable(Str), null],
            touch_action: [Nullable(Str), null],
            transform: [Nullable(Str), null],
            transform_box: [Nullable(Str), null],
            transform_origin: [Nullable(Str), null],
            transform_style: [Nullable(Str), null],
            transition: [Nullable(Str), null],
            transition_delay: [Nullable(Str), null],
            transition_duration: [Nullable(Str), null],
            transition_property: [Nullable(Str), null],
            transition_timing_function: [Nullable(Str), null],
            translate: [Nullable(Str), null],
            unicode_bidi: [Nullable(Str), null],
            user_select: [Nullable(Str), null],
            vertical_align: [Nullable(Str), null],
            visibility: [Nullable(Str), null],
            white_space: [Nullable(Str), null],
            widows: [Nullable(Str), null],
            width: [Nullable(Str), null],
            will_change: [Nullable(Str), null],
            word_break: [Nullable(Str), null],
            word_spacing: [Nullable(Str), null],
            word_wrap: [Nullable(Str), null],
            writing_mode: [Nullable(Str), null],
            z_index: [Nullable(Str), null],
        }));
    }
}
